File: src/main/java/com/stolser/javatraining/webproject/controller/form/validator/user/UserEmailValidator_.java
package com.stolser.javatraining.webproject.controller.form.validator.user;

import com.stolser.javatraining.webproject.controller.form.validator.AbstractValidator;
import com.stolser.javatraining.webproject.controller.form.validator.ValidationResult;
import com.stolser.javatraining.webproject.service.impl.UserServiceImpl;

import javax.servlet.http.HttpServletRequest;
import java.util.Optional;
import java.util.regex.Pattern;

import static com.stolser.javatraining.webproject.controller.ApplicationResources.*;

public class UserEmailValidator_ extends AbstractValidator {
    private static ValidationResult regexFailedResult =
            new ValidationResult(STATUS_CODE_VALIDATION_FAILED, MSG_USER_EMAIL_REGEX_ERROR);
    private static ValidationResult duplicationFailedResult =
            new ValidationResult(STATUS_CODE_VALIDATION_FAILED, MSG_USER_EMAIL_DUPLICATION_ERROR);

    private UserEmailValidator_() {}

    private static class InstanceHolder {
        private static final UserEmailValidator_ INSTANCE = new UserEmailValidator_();
    }

    public static UserEmailValidator_ getInstance() {
        return InstanceHolder.INSTANCE;
    }

    @Override
    public Optional<ValidationResult> checkParameter(String userEmail, HttpServletRequest request) {
        if (!emailMatchesRegex((userEmail))) {
            return Optional.of(regexFailedResult);
        }

        if (emailExistsInDb(userEmail)) {
            return Optional.of(duplicationFailedResult);
        }

        return Optional.empty();
    }

    private boolean emailMatchesRegex(String userEmail) {
        return Pattern.matches(USER_EMAIL_PATTERN_REGEX, userEmail);
    }

    private boolean emailExistsInDb(String userEmail) {
        return UserServiceImpl.getInstance().emailExistsInDb(userEmail);
    }
}

File: src/main/scala/com/stolser/javatraining/webproject/controller/form/validator/user/UserEmailValidator.scala
package com.stolser.javatraining.webproject.controller.form.validator.user

import com.stolser.javatraining.webproject.controller.ApplicationResources.{MSG_USER_EMAIL_DUPLICATION_ERROR, MSG_USER_EMAIL_REGEX_ERROR, STATUS_CODE_VALIDATION_FAILED, USER_EMAIL_PATTERN_REGEX}
import com.stolser.javatraining.webproject.controller.form.validator.{AbstractValidator, ValidationResult}
import com.stolser.javatraining.webproject.service.UserService
import com.stolser.javatraining.webproject.service.impl.mysql.UserServiceMysqlImpl
import javax.servlet.http.HttpServletRequest

/**
	* Created by Oleg Stoliarov on 10/7/18.
	*/
object UserEmailValidator extends AbstractValidator {
	private var userServiceImpl = UserServiceMysqlImpl: UserService
	private val regexFailedResult = ValidationResult(
		statusCode = STATUS_CODE_VALIDATION_FAILED,
		messageKey = MSG_USER_EMAIL_REGEX_ERROR
	)
	private val duplicationFailedResult = ValidationResult(
		statusCode = STATUS_CODE_VALIDATION_FAILED,
		messageKey = MSG_USER_EMAIL_DUPLICATION_ERROR
	)

	override protected def checkParameter(userEmail: String,
																				request: HttpServletRequest): Option[ValidationResult] = {
		if (emailIsNotValid(userEmail)) return Some(regexFailedResult)
		if (emailExistsInDb(userEmail)) return Some(duplicationFailedResult)

		Option.empty[ValidationResult]
	}

	private def emailIsNotValid(userEmail: String) = !userEmail.matches(USER_EMAIL_PATTERN_REGEX)

	private def emailExistsInDb(userEmail: String) = userServiceImpl.emailExistsInDb(userEmail)

	private[user] def userService: UserService = userServiceImpl

	private[user] def userService_=(userService: UserService): Unit = {
		require(userService != null)

		this.userServiceImpl = userService
	}
}
